### ISSUE TITLE
PR 4 — `InferenceLayer` base + `SingleInstanceLayer`

### ISSUE THREAD

**Issue by gitttt-1234** [talmolab/sleap-nn#512]
**Parent epic:** #508
**Prerequisites:** #509, #510, #511
**Blocks:** #513, #514, #515

---

## Goal

Introduce the `InferenceLayer` ABC and ship `SingleInstanceLayer` as the proof-of-pattern. After this PR, the end-to-end path `numpy image → Outputs` works for single-instance models using the new architecture, side-by-side with the existing code.

Single-instance is the right first layer because it's the simplest — global peak per channel, no crops, no PAF grouping, no class decoding. If the pattern works here cleanly, #514 applies it to the other five layer types mechanically.

### `PreprocessConfig` + `PostprocessConfig`

```python
# sleap_nn/inference/layers/base.py  (config portion)

from typing import List, Literal, Optional, Tuple
import attrs


@attrs.define
class PreprocessConfig:
    """Controls how raw input images are transformed before the model."""
    input_scaling: float = 1.0
    pad_to_stride: int = 32
    ensure_rgb: bool = False
    ensure_grayscale: bool = False
    max_height: Optional[int] = None
    max_width: Optional[int] = None
    crop_size: Optional[Tuple[int, int]] = None
    anchor_part: Optional[str] = None
    normalize: bool = True


@attrs.define
class PAFConfig:
    """Bottom-up part-affinity-field grouping parameters."""
    max_edge_length_ratio: float = 0.25
    dist_penalty_weight: float = 1.0
    n_points: int = 10
    min_instance_peaks: int = 0
    min_line_scores: float = 0.25
    max_peaks_per_node: Optional[int] = None


@attrs.define
class PostprocessConfig:
    """Controls how raw model outputs become keypoint predictions."""
    peak_threshold: float = 0.2
    refinement: Literal["integral", "local", "none"] = "integral"
    integral_patch_size: int = 5
    max_instances: Optional[int] = None
    paf: Optional[PAFConfig] = None

    return_confmaps: bool = False
    return_pafs: bool = False
    return_paf_graph: bool = False
    return_class_maps: bool = False
    return_class_vectors: bool = False
```

### `InferenceLayer` base class

```python
# sleap_nn/inference/layers/base.py  (layer portion)

from abc import ABC, abstractmethod
from typing import Any, Dict, List, Optional, Tuple, Union
import attrs
import numpy as np
import torch
import sleap_io as sio

from sleap_nn.inference.outputs import Outputs
from sleap_nn.inference.layers.backends.base import ModelBackend


@attrs.define
class InferenceLayer(ABC):
    """Base class for all model-type-specific inference layers.

    Owns the complete single-frame pipeline:
        raw input -> preprocess -> backend forward -> postprocess -> Outputs

    Subclasses override postprocess() with model-type-specific logic.
    The backend is injected and handles runtime differences.
    """

    backend: ModelBackend
    skeleton: sio.Skeleton
    preprocess_config: PreprocessConfig = attrs.Factory(PreprocessConfig)
    postprocess_config: PostprocessConfig = attrs.Factory(PostprocessConfig)
    output_stride: int = 1
    input_channels: int = 1

    # ── Main entry point ──────────────────────────────────────────────

    def predict(
        self, x: Union[np.ndarray, torch.Tensor, List[np.ndarray]],
    ) -> Outputs:
        """Full pipeline: preprocess -> forward -> postprocess."""
        processed, info = self.preprocess(x)
        raw_out = self.run(processed)
        return self.postprocess(raw_out, info)

    # ── Preprocessing (shared by all model types) ─────────────────────

    def preprocess(
        self, imgs: Union[np.ndarray, torch.Tensor, List[np.ndarray]],
    ) -> Tuple[torch.Tensor, PreprocInfo]:
        """Normalize shape, convert channels, resize, pad, transfer to device.

        Steps (per §4.4):
          1. Shape normalize to (B, C, H, W) uint8
          2. Channel convert (ensure_rgb | ensure_grayscale)
          3. Size match to (max_height, max_width) preserving AR → eff_scale
          4. Input-scale resize (if input_scale != 1.0)
          5. Pad to stride
          6. DEFER normalization (stay uint8 on CPU — 4× PCIe savings)
          7. Move to device
          8. Normalize on device (uint8 → float / 255)
        """
        ...

    # ── Backend dispatch ──────────────────────────────────────────────

    def run(self, x: torch.Tensor) -> Dict[str, torch.Tensor]:
        return self.backend(x)

    # ── Postprocessing (override per model type) ──────────────────────

    @abstractmethod
    def postprocess(
        self,
        raw_outputs: Dict[str, torch.Tensor],
        info: PreprocInfo,
    ) -> Outputs:
        ...

    # ── Coordinate transform helpers ──────────────────────────────────

    def _reverse_coords(self, coords: torch.Tensor, info: PreprocInfo) -> torch.Tensor:
        """Apply: coords * output_stride / input_scale / eff_scale"""
        from sleap_nn.inference.ops.coord import (
            undo_stride, undo_input_scale, undo_eff_scale,
        )
        coords = undo_stride(coords, info.output_stride)
        coords = undo_input_scale(coords, info.input_scale)
        coords = undo_eff_scale(coords, info.eff_scale)
        return coords

    def warmup(self, sample_shape: Optional[Tuple[int, ...]] = None) -> None:
        if sample_shape is None:
            h = self.preprocess_config.max_height or 384
            w = self.preprocess_config.max_width or 384
            sample_shape = (1, self.input_channels, h, w)
        self.backend.warmup(sample_shape)

    @classmethod
    def from_checkpoint(
        cls,
        checkpoint_path: str,
        device: str = "auto",
        use_compile: bool = False,
        use_fp16: bool = False,
        **kwargs,
    ) -> "InferenceLayer":
        """Load from a sleap-nn checkpoint directory.

        Auto-detects model type and returns the right subclass.
        """
        from sleap_nn.inference.layers.backends.loader import load_layer
        return load_layer(checkpoint_path, device=device,
                          use_compile=use_compile, use_fp16=use_fp16, **kwargs)
```

### `SingleInstanceLayer`

```python
# sleap_nn/inference/layers/single_instance.py

from typing import Dict
import attrs
import torch

from sleap_nn.inference.layers.base import InferenceLayer, PreprocInfo
from sleap_nn.inference.ops.peaks import find_global_peaks
from sleap_nn.inference.outputs import Outputs


@attrs.define
class SingleInstanceLayer(InferenceLayer):
    """Single-instance model: one pose per frame.

    Postprocessing: global peak finding (one peak per node channel).
    """

    def postprocess(
        self, raw_out: Dict[str, torch.Tensor], info: PreprocInfo,
    ) -> Outputs:
        if self.backend.does_baked_postproc:
            peaks = raw_out["peaks"]             # (B, N, 2)
            peak_vals = raw_out["peak_vals"]     # (B, N)
        else:
            confmaps = raw_out.get(
                "SingleInstanceConfmapsHead",
                next(iter(raw_out.values())),
            )
            peaks, peak_vals = find_global_peaks(
                confmaps,
                threshold=self.postprocess_config.peak_threshold,
                refinement=self.postprocess_config.refinement,
                integral_patch_size=self.postprocess_config.integral_patch_size,
            )

        peaks = self._reverse_coords(peaks, info)

        return Outputs(
            pred_keypoints=peaks.unsqueeze(1),          # (B, N, 2) → (B, 1, N, 2)
            pred_peak_values=peak_vals.unsqueeze(1),
            instance_scores=peak_vals.mean(dim=-1, keepdim=True),
            pred_confmaps=(
                raw_out.get("SingleInstanceConfmapsHead")
                if self.postprocess_config.return_confmaps else None
            ),
            preprocess_info=info,
        )
```

### Backend loader (minimal, first pass)

```python
# sleap_nn/inference/layers/backends/loader.py

def load_layer(checkpoint_path, device="auto", **kwargs):
    """Load an InferenceLayer from a checkpoint directory.

    Dispatches on:
    - Config file: training_config.yaml → modern, training_config.json → legacy (#520)
    - Head type in config → which InferenceLayer subclass to instantiate
    - File extension: .ckpt → TorchBackend, .onnx/.trt → #515
    """
    ...
```

For #512, only the single-instance + modern-config + Torch path needs to work. The rest can raise `NotImplementedError("PR <N>")`.

## Parity testing

The critical gate for this PR: outputs must match `SingleInstanceInferenceModel` numerically on fixture inputs.

```python
# tests/inference/layers/test_single_instance_layer.py

def test_single_instance_parity():
    ckpt = "tests/assets/minimal_instance/single_instance/best.ckpt"
    image = load_test_image()

    # Old path
    old_model = load_trained_model_from_ckpt(ckpt)
    old_output = old_model({"image": image})

    # New path
    layer = SingleInstanceLayer.from_checkpoint(ckpt, device="cpu")
    new_output = layer.predict(image)

    torch.testing.assert_close(
        new_output.pred_keypoints,
        old_output["pred_instance_peaks"],
        rtol=1e-5, atol=1e-5,
    )
    torch.testing.assert_close(
        new_output.pred_peak_values,
        old_output["pred_peak_values"],
        rtol=1e-5, atol=1e-5,
    )
```

## Acceptance criteria

- [ ] `InferenceLayer` ABC with all methods from design doc §4B.4
- [ ] `SingleInstanceLayer.from_checkpoint()` works on fixture `single_instance/best.ckpt`
- [ ] Numpy input → Outputs: `layer.predict(np.ndarray) → Outputs` works end-to-end
- [ ] Parity test passes (byte-for-byte against `SingleInstanceInferenceModel`)
- [ ] `warmup()` works on CPU/MPS without error
- [ ] `PreprocInfo.eff_scale` roundtrips correctly through preprocess → postprocess (regression test for the #1 miscalibration risk called out in §4.5)
- [ ] `ruff check` + `black --check` clean

## Files touched

**Created:**
- `sleap_nn/inference/layers/base.py` (extended from #510)
- `sleap_nn/inference/layers/single_instance.py`
- `sleap_nn/inference/layers/backends/loader.py` (minimal — only single-instance + torch path)
- `tests/inference/layers/test_single_instance_layer.py`
- `tests/inference/layers/test_preprocess.py` (covers the preprocess path)

## Out of scope

- Other model types (#514)
- ONNX/TensorRT backends (#515)
- Legacy SLEAP ≤1.4 loader (#520)

**Comment by gitttt-1234** [comment]
Closed — merged as part of the inference refactor stack (epic #508).
